### PR TITLE
TINKERPOP3-886

### PR DESCRIPTION
**Pull request by kushal256**
Let me know if exceptions should be managed any differently, or if you see any issues.

### CODE REVIEW

File: docs/src/implementations.asciidoc [hunk 1]
```diff
@@ -630,7 +630,8 @@ TinkerGraph has several settings that can be provided on creation via `Configura
 value is specified here, the the `gremlin.tinkergraph.graphFormat` should also be specified.  If this value is not
 included (default), then the graph will stay in-memory and not be loaded/persisted to disk.
 |gremlin.tinkergraph.graphFormat |The format to use to serialize the graph which may be one of the following:
-`graphml`, `graphson`, or `gryo`. If a value is specified here, the the `gremlin.tinkergraph.graphLocation` should
+`graphml`, `graphson`, gryo`, or a fully qualified class name that implements Io.Builder interface.
+If a value is specified here, the the `gremlin.tinkergraph.graphLocation` should
```
Comment (dkuppitz) [inline docs/src/implementations.asciidoc]: Typo."the the" => "then the"

**Comment by spmallette** [comment]
Thanks for working on this. I see you did something for reading a graph, but what about writing?

https://github.com/kushal256/incubator-tinkerpop/blob/master/tinkergraph-gremlin/src/main/java/org/apache/tinkerpop/gremlin/tinkergraph/structure/TinkerGraph.java#L296

Note that the docs needed to be updated:

https://github.com/apache/incubator-tinkerpop/blob/master/docs/src/implementations.asciidoc#configuration

Specifically, the configuration description for `gremlin.tinkergraph.graphFormat` needs something to denote this new feature.  

Would be nice to see a test case as well.  Perhaps you could stub a reader/writer and assert that the class gets created and the appropriate methods are called. wdyt?

I'm not sure how others  feel, but I think we will wait to merge this until after release of 3.1.0-incubating (in code freeze for that now).

**Comment by dkuppitz** [comment]
I agree, I'd also like to see that covered by one or two tests.

**Comment by kushal256** [comment]
Thanks for your comments.  Yes, there should be similar logic for writing
graph.

Also, I agree, there should be tests, but I'm not sure how to do so when
using reflection (Class.forName() and newInstance() for creating new
Io.Builder), with mockito's stubs.  Any recommendations for how to do this?

**Comment by spmallette** [comment]
Don't use mockito.  Create your mock the old fashioned way and just create a concrete class in test that implements the interfaces you need.

**Comment by kushal256** [comment]
OK just checked in with an old fashioned mock.

File: tinkergraph-gremlin/src/test/java/org/apache/tinkerpop/gremlin/tinkergraph/structure/TinkerGraphTest.java [hunk 2]
```diff
@@ -23,19 +23,21 @@
...
 import org.apache.tinkerpop.gremlin.structure.io.IoCore;
+import org.apache.tinkerpop.gremlin.structure.io.IoRegistry;
 import org.apache.tinkerpop.gremlin.structure.io.IoTest;
 import org.junit.Ignore;
 import org.junit.Test;
 
-import java.io.ByteArrayInputStream;
-import java.io.ByteArrayOutputStream;
-import java.io.File;
+import java.io.*;
```
Comment (spmallette) [inline tinkergraph-gremlin/src/test/java/org/apache/tinkerpop/gremlin/tinkergraph/structure/TinkerGraphTest.java]: not a big deal, but for future pull requests, note that for style purposes we don't use wildcard imports.

**Comment by spmallette** [comment]
thanks for updating the PR.  we're in the midst of releasing 3.1.0-incubating.  we'll review this in more detail as soon as that is out the door.

**Comment by spmallette** [comment]
I think we're good to VOTE on this now.  Ran `mvn clean install` and all tests pass.  There's a few little odds/ends I'll cleanup after merge.

+1

**Comment by twilmes** [comment]
Looks good to me.  Ran the tests and then played around with it in the shell a bit.  Exception message is good on a bogus persistence engine and gryo and graphson work as expected.

+1

**Comment by okram** [comment]
I pulled the branch and `mvn clean install`'d. Everything passed. I also reviewed the code. I don't really know this area of the codebase but the code is in our style and Stephen agrees with the PR, so thats good enough for me.

VOTE +1

**Comment by dkuppitz** [comment]
I've been almost happy with the first version and now that something has been added to the docs:

VOTE: +1

**Comment by kushal256** [comment]
Awesome!  Apologies for the import with a wildcard, my Intellij did that
for me, I can clean that up, let me know if there are any additional
cleanup items, want to ensure everything meets the project's standards.

**Comment by spmallette** [comment]
@kushal256 not a problem - i figured your IDE was the culprit.  if you'd like to amend your PR with the fix to the imports that will be fine.  

The other thing we tend to do is mark all variables `final` (unless they are not - which really makes them jump out at you and makes you wonder why they are not when you see them in the code base). so for example you added:

```text
    public static Io.Builder createIoBuilder(String graphFormat) throws ClassNotFoundException, IllegalAccessException, InstantiationException {
        Class<Io.Builder> ioBuilderClass = (Class<Io.Builder>) Class.forName(graphFormat);
        Io.Builder ioBuilder = ioBuilderClass.newInstance();
        return ioBuilder;
    }
```

`graphFormat`, `ioBuilder` and `ioBuilderClass` should be `final`.  `IoCore` is also a "public" API class, so some javadoc on that method would be good.  

And last but not least, if you want to really tie a bow on this, you would add an entry to the CHANGELOG for 3.1.1:

https://github.com/apache/incubator-tinkerpop/blob/master/CHANGELOG.asciidoc#tinkerpop-311-not-officially-released-yet

Please comment when you've updated the PR with those little changes and i'll merge after that.

**Comment by spmallette** [comment]
@kushal256 i see you made some changes - is this ready for me to merge now?

**Comment by kushal256** [comment]
@spmallette, yes, I made the changes as per your recommendation, ready for
merge.

@kushal256 <https://github.com/kushal256> i see you made some changes - is
this ready for me to merge now?

—
Reply to this email directly or view it on GitHub
<https://github.com/apache/incubator-tinkerpop/pull/141#issuecomment-158980645>
.

**Comment by spmallette** [comment]
@kushal256 i've merged this one.  thanks for working through all that.  note that were some weird file permission changes that came with your PR.  I reverted those:

https://github.com/apache/incubator-tinkerpop/commit/5f1a354e577b7c31f00469c91acae61a6144a85f

for some reason, all those files were made executable.  ???  anyway, something to watch for your next PR.  thanks again!

**Pull request merged**